### PR TITLE
Move volume dynamic provisioning scheduling to beta

### PR DESCRIPTION
**Pull request by lichuqiang**
**What this PR does / why we need it**:

*  Combine feature gate VolumeScheduling and DynamicProvisioningScheduling into one
* Add allowedTopologies description in kubectl

**Special notes for your reviewer**:
Wait until related e2e and downside plugins are ready.

/hold

**Release note**:

```release-note
Move volume dynamic provisioning scheduling to beta (ACTION REQUIRED: The DynamicProvisioningScheduling alpha feature gate has been removed. The VolumeScheduling beta feature gate is still required for this feature)
```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/pv_controller.go
Comment (msau42) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: The annotation needs to move to beta too.  I would actually not prefix the annotation with "beta" either, it will make going to GA easier.

File: pkg/printers/internalversion/describe.go
Comment (msau42) [inline pkg/printers/internalversion/describe.go]: Can you paste an example of the output?

Comment (lichuqiang) [reply pkg/printers/internalversion/describe.go]: out

Similar to that of node affinity

**Comment by msau42** [comment]
/lgtm

We need to wait for these PRs to go in first:
https://github.com/kubernetes/kubernetes/pull/66843
https://github.com/kubernetes/kubernetes/pull/67530

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go]: shouldn't this change to true and beta, not be removed completely?

Comment (msau42) [reply pkg/features/kube_features.go]: We want to merge this with the VolumeScheduling feature gate, which is already beta

Comment (msau42) [reply pkg/features/kube_features.go]: According to the feature gate discussion [here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-architecture/qJeJ7ANYzS0/qll1H4GoDQAJ), when a flag goes GA, we have to deprecate it before removing, to support upgrade cases.  

However, this is an alpha feature gate, so we don't support upgrade.  I think it is fine to remove/merge it in this case.  @liggitt wdyt?

**Comment by msau42** [comment]
The other two prs have merged, so this can go in.
/hold cancel

File: pkg/features/kube_features.go
Comment (liggitt) [reply pkg/features/kube_features.go]: if the gate never graduated from alpha, we can drop it on a release boundary (though it's release-note worthy).

based on offline discussion, the code gated by DynamicProvisioningScheduling is also opt-in based on storageclass attributes. if that's the case, I see no downside to collapsing onto the VolumeScheduling

Comment (msau42) [reply pkg/features/kube_features.go]: Making sure all entry points to the feature are covered by:
1) VolumeScheduling feature gate
2) Opt-in behavior via StorageClass.VolumeBindingMode field

Scheduler:
- Feature gate check is at the beginning of the [predicate](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L1675)
- In predicate call, we check for delayed binding setting and treat it as [legacy](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/scheduler_binder.go#L383) (immediate binding) scenario otherwise
- Assume/bind volumes are feature [gated](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go#L268)

PV controller:
- Defer to the scheduler is controlled by feature [gate](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L289) and StorageClass [check](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L317)

API server:
- StorageClass field controlled by feature [gate](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/storage/validation/validation.go#L239)

Comment (msau42) [reply pkg/features/kube_features.go]: @lichuqiang can you update the release note saying that the feature gate has been merged with VolumeScheduling.

Comment (lichuqiang) [reply pkg/features/kube_features.go]: done

Comment (msau42) [reply pkg/features/kube_features.go]: sorry being nitpicky, can you word it as "ACTION REQUIRED: The DynamicProvisioningScheduling alpha feature gate has been removed.  The VolumeScheduling beta feature gate is still required for this feature"

Comment (lichuqiang) [reply pkg/features/kube_features.go]: sure thing

**Comment by msau42** [comment]
/lgtm

/assign @saad-ali @liggitt 
for approval

**Review by saad-ali — APPROVED** [review]
/approve
/milestone v1.12

**Pull request merged**
